services/core/shared/core/compat/sdk_v5/blocks.js

```diff
@@ -417,95 +411,57 @@ const cacheLegacyAccountInfo = async () => {
 
-					// Index the genesis block accounts
-					const initDelegateAddresses = genesisBlock.asset.initDelegates;
-					await indexAccountsbyAddress(initDelegateAddresses, true);
-				}
+		if (pageNum >= lastCachedPage) {
+			const genesisAccountAddressesToIndex = genesisAccountsToIndex.slice(currentPage, nextPage);
 
-				logger.info(`Indexing genesis account batch: ${batchNum}, ${Math.ceil(genesisAccountsToIndex.length / BATCH_SIZE) - batchNum - 1} to go`);
+			logger.info(`Scheduling retrieval of genesis accounts batch ${pageNum + 1}/${NUM_PAGES} (${percentage}%)`);
 
-				const PAGE_SIZE = 1000;
-				const NUM_PAGES = Math.ceil(genesisAccountAddressesToIndex.length / PAGE_SIZE);
-				for (let i = 0; i < NUM_PAGES; i++) {
-					// eslint-disable-next-line no-await-in-loop
-					await indexAccountsbyAddress(
-						genesisAccountAddressesToIndex.slice(i * PAGE_SIZE, (i + 1) * PAGE_SIZE),
-						true,
-					);
-				}
+			/* eslint-disable no-await-in-loop */
+			await indexAccountsbyAddress(genesisAccountAddressesToIndex, true);
+			await genesisAccountsCache.set(genesisAccountPageCached, pageNum);
+			/* eslint-enable no-await-in-loop */
+		} else {
+			logger.info(`Skipping retrieval of genesis accounts batch ${pageNum + 1}/${NUM_PAGES} (${percentage}%)`);
+		}
+	}
+};
 
-				if (genesisAccountAddressesToIndex.length < BATCH_SIZE) {
-					// Stop retrying genesis account indexing on successful completion
-					Signals.get('newBlock').remove(indexGenesisAccounts);
-					logger.info('Finished indexing the genesis block accounts');
+const indexGenesisAccounts = async () => {
+	const isScheduled = await genesisAccountsCache.get('isGenesisAccountIndexingScheduled');
 
-					// Reset global variables and free memory
-					genesisAccountIndexingBatchNum = -1;
-					genesisAccountsToIndex = undefined;
-				}
-			}
+	if (isScheduled === true) {
+		logger.info('Skipping genesis account index update (one-time operation, already indexed)');
+		return;
+	}
 
-			// On success, update the cache with the batch number
-			await genesisAccountsCache.set('indexedTillBatch', genesisAccountIndexingBatchNum);
-		}
+	try {
+		logger.info('Attempting to update genesis account index (one-time operation)');
+		await performGenesisAccountsIndexing();
+		await genesisAccountsCache.set('isGenesisAccountIndexingScheduled', true);
+		await genesisAccountsCache.set('genesisAccountPageCached', 0);
 	} catch (err) {
-		logger.warn(`Unable to index Genesis block accounts batch ${genesisAccountIndexingBatchNum}, will retry again: ${err}`);
-		genesisAccountIndexingBatchNum--;
-		await genesisAccountsCache.set('indexedTillBatch', genesisAccountIndexingBatchNum);
-	} finally {
-		if (numAccountsIndexed >= BATCH_SIZE * (genesisAccountIndexingBatchNum + 1)) {
-			isGenesisAccountsIndexingInProgress = false;
-		}
+		logger.error('Critical error: Unable to index Genesis block accounts batch. Will retry after the restart');
+		logger.error(err.message);
+		process.exit(1);
```

Comment:
Since we are terminating the application prematurely, should we not mark these logs `FATAL`?
```suggestion
		logger.fatal('Critical error: Unable to index Genesis block accounts batch. Will retry after the restart');
		logger.fatal(err.message);
		process.exit(1);
```

Reply:
Good point! Thanks @sameersubudhi